src/scheduler/multi_threading.c

```diff
@@ -0,0 +1,225 @@
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <pthread.h>
+
+#include "log.h"
+#include "avltree.h"
+
+#include "misc.h"
+#include "node_info.h"
+#include "data_types.h"
+#include "globals.h"
+#include "node_info.h"
+#include "queue.h"
+#include "fifo.h"
+#include "resource_resv.h"
+
+void *worker(void* arg);
+
+
+/**
+ * @brief	create the thread id key & set it for the main thread
+ *
+ * @param	void
+ *
+ * @return	void
+ */
+static void
+create_id_key(void)
+{
+	int *mainid;
+
+	mainid = malloc(sizeof(int));
+	if (mainid == NULL) {
+		snprintf(log_buffer, sizeof(log_buffer), "Malloc error");
+		schdlog(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, log_buffer);
+		return;
+	}
+	*mainid = 0;
+
+	pthread_key_create(&th_id_key, free);
+	pthread_setspecific(th_id_key, (void *) mainid);
+}
+
+
+/**
+ * @brief	initialize multi-threading
+ *
+ * @param	void
+ *
+ * @return	int
+ * @retval	0 for success
+ * @retval	1 for malloc error
+ */
+int
+init_multi_threading(void)
+{
+	if (threads == NULL) {
+		int i;
+		int num_cores;
+		pthread_mutexattr_t attr;
+
+		/* Create task and result queues */
+		work_queue = new_schd_queue();
+		result_queue = new_schd_queue();
+		if (work_queue == NULL || result_queue == NULL) {
+			snprintf(log_buffer, sizeof(log_buffer), "Error malloc'ing thread memory");
+			schdlog(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, log_buffer);
+			return 1;
+		}
+
+		threads_die = 0;
+		if (pthread_cond_init(&work_cond, NULL) != 0) {
+			snprintf(log_buffer, sizeof(log_buffer), "pthread_cond_init failed");
+			schdlog(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, log_buffer);
+			return 1;
+		}
+		if (pthread_cond_init(&result_cond, NULL) != 0) {
+			snprintf(log_buffer, sizeof(log_buffer), "pthread_cond_init failed");
+			schdlog(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, log_buffer);
+			return 1;
+		}
+
+		if (pthread_mutexattr_init(&attr) != 0) {
+			snprintf(log_buffer, sizeof(log_buffer), "pthread_mutexattr_init failed");
+			schdlog(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, log_buffer);
+			return 1;
+		}
+		if (pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE)) {
+			snprintf(log_buffer, sizeof(log_buffer), "pthread_mutexattr_settype failed");
+			schdlog(PBSEVENT_ERROR, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__, log_buffer);
+			return 1;
+		}
+		pthread_mutex_init(&work_lock, &attr);
+		pthread_mutex_init(&result_lock, &attr);
+
+		/* Create as many threads as the number of cores */
+#ifdef WIN32
+		SYSTEM_INFO sysinfo;
+		GetSystemInfo(&sysinfo);
+		num_cores = sysinfo.dwNumberOfProcessors;
+#else
+		num_cores = sysconf(_SC_NPROCESSORS_ONLN);
+#endif
```

Comment:
Hmm .. I see your point. A few things:
- the threads sleep until work is available for them, they aren't spin looping, so cores won't be consumed until needed.
- the work is segregated in bug chunks for the worker threads. So, if a site has less than 1024 nodes, only 1 worker thread will be woken up for the node functions. Similarly, if there are less than 1024 jobs in the queue at a time, only 1 worker thread will be woken up for the job functions. So, unless a site has a large number of jobs/nodes, not all worker threads will be used.
- In my testing with a million jobs and 50k nodes on a machine with 8vcpus (4 cores, 2 hyperthreads each), I still didn't see the overall CPU% go beyond 400%.

So, we can make this a configurable, but I don't like the idea of using just half the cores by default. Maybe we should use N - 1 cores? leave one for the server since it's single threaded? @billnitzberg seeking advice from you as well, thanks!

Reply:
The problem with using N-1 is that we're planning on horizontally scaling multiple PBS components (e.g. multi-server).  We don't want to hog all the CPUs for the scheduler, or worse, have the PBS components fight for CPU cycles and slow things down.  We usually put multiple schedulers on one machine (they don't have to), so using all of the CPUs will clash with different schedulers.  I definitely think we should make it configurable.  I'm open to ideas about the default though.  I am curious what @billnitzberg and @subhasisb has to say about this.

Reply:
A couple ideas:
(1) Make it configurable & set the default = 1. This allows rolling out this (risky) change slowly, only to those who explicitly set a value > 1. We can change the default in a future release.
(2) Benchmark a handful of representative workloads on a machine with lots (>20) cores, varying the numbers of threads, and see where the knee in the curve is and use that as the default. I suspect somewhere between 2 - 20 threads will be a good default.
(3) What does the server do for TPP threads? What about Postgresql? Both these will likely run on the same node as the Scheduler.

Reply:
I like your first approach although I think the default should not be 1. Making it 1 will make it slower than what it is today. Maybe, by default, we don't spawn any threads at all and keep the behavior as it is today.